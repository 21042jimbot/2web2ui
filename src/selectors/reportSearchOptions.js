import { createSelector } from 'reselect';
import moment from 'moment';
import { stringifyTypeaheadfilter } from 'src/helpers/string';
import _ from 'lodash';

const selectDateOptions = state => ({
  from: moment(state.reportOptions.from).format(),
  to: moment(state.reportOptions.to).format(),
  precision: state.reportOptions.precision,
  timezone: state.reportOptions.timezone,
  range: state.reportOptions.relativeRange,
});

const selectTypeaheadFilters = state => ({
  filters: _.get(state, 'reportOptions.filters', []).map(stringifyTypeaheadfilter),
});

const selectSummaryMetrics = state => ({
  metrics: _.get(state, 'reportOptions.metrics', []).map(metric =>
    typeof metric === 'string' ? metric : metric.key,
  ),
});

/**
 * Converts reportOptions for url sharing
 */
export const selectReportSearchOptions = createSelector(
  [selectDateOptions, selectTypeaheadFilters],
  (dates, filters) => ({ ...dates, ...filters }),
);

/**
 * Converts reportOptions for url sharing for the summary chart
 */
export const selectSummaryChartSearchOptions = createSelector(
  [selectDateOptions, selectTypeaheadFilters, selectSummaryMetrics],
  (dates, filters, metrics) => ({ ...dates, ...filters, ...metrics }),
);
