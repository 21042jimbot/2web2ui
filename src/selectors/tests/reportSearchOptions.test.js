import { selectReportSearchOptions, selectSummaryChartSearchOptions } from '../reportSearchOptions';

jest.mock('src/helpers/string', () => ({
  stringifyTypeaheadfilter: jest.fn(filter => filter.id),
}));

describe('ReportSearchOptions Selectors', () => {
  let state;
  beforeEach(() => {
    state = {
      reportOptions: {
        from: '2018-03-23T17:10:08-04:00',
        to: '2018-03-23T17:11:08-04:00',
        range: null,
        filters: [{ id: 101 }, { id: 102 }],
        metrics: ['count_bounce', 'count_accepted', { key: 'metric_key' }],
        precision: 'hour',
        timezone: 'America/New_York',
      },
    };
  });

  describe('selectReportSearchOptions', () => {
    it('prepares reportOptions for URL sharing', () => {
      expect(selectReportSearchOptions(state)).toMatchSnapshot();
    });
  });

  describe('selectSummaryChartSearchOptions', () => {
    it('prepares summary chart reportOptions for URL sharing', () => {
      expect(selectSummaryChartSearchOptions(state)).toMatchSnapshot();
    });
  });
});
