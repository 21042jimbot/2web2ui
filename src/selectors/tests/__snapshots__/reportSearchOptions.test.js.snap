// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ReportSearchOptions Selectors selectReportSearchOptions prepares reportOptions for URL sharing 1`] = `
Object {
  "filters": Array [
    101,
    102,
  ],
  "from": "2018-03-23T21:10:08Z",
  "range": undefined,
  "to": "2018-03-23T21:11:08Z",
}
`;

exports[`ReportSearchOptions Selectors selectSummaryChartSearchOptions prepares summary chart reportOptions for URL sharing 1`] = `
Object {
  "filters": Array [
    101,
    102,
  ],
  "from": "2018-03-23T21:10:08Z",
  "metrics": Array [
    "count_bounce",
    "count_accepted",
    "metric_key",
  ],
  "range": undefined,
  "to": "2018-03-23T21:11:08Z",
}
`;
